### PR TITLE
Add panning in price chart

### LINKED ISSUE #5: Support panning in price chart
Use left / right arrow keys to move backwards / forwards in time, while keeping the same time frame.

### PR DESCRIPTION
**Pull request by leesiongchan**
Fix #5

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -169,6 +171,20 @@ pub enum TimeFrame {
 }
 
 impl TimeFrame {
+    pub fn duration(self) -> Option<Duration> {
```
Comment (leesiongchan) [inline src/app.rs:174]: This part suck, do you have any idea?

File: src/ui.rs [hunk 2]
```diff
@@ -85,7 +85,13 @@ fn draw_header<B: Backend>(
         .insert(UiTarget::StockName, stock_name_area);
 }
 
-fn draw_body<B: Backend>(f: &mut Frame<B>, App { stock, .. }: &App, area: Rect) {
+fn draw_body<B: Backend>(
+    f: &mut Frame<B>,
+    App {
+        stock, ui_state, ..
```
Comment (leesiongchan) [inline src/ui.rs:91]: Why do we need a `..` here?

File: src/app.rs (hunk 1 above)
Comment (teohhanhui) [reply src/app.rs:174]: Why does it suck? This is the idiomatic way of doing it in Rust, as far as I can tell. You can't easily attach constant properties to enum variants... I mean there's https://docs.rs/strum/0.18.0/strum/trait.EnumProperty.html but...

Comment (leesiongchan) [reply src/app.rs:174]: I mean the way I calc the duration is inaccurate.

File: src/app.rs [hunk 3]
```diff
@@ -16,6 +17,7 @@ pub struct App {
 
 #[derive(Debug)]
 pub struct UiState {
+    pub end_date: DateTime<Utc>,
```
Comment (teohhanhui) [inline src/app.rs:20]: Hmm... Maybe we can use `start_date` instead? WDYT? I'm not sure which makes more sense. This probably requires some domain knowledge... I'll try to read up.

File: src/app.rs (hunk 1 above)
Comment (leesiongchan) [reply src/app.rs:174]: I consider 30 days = 1 month, do we have better way of doing?

File: src/app.rs (hunk 3 above)
Comment (teohhanhui) [reply src/app.rs:20]: Actually, maybe we should have both, no? We might want to support custom start / end date in the future. Not sure... Haha.

Comment (teohhanhui) [reply src/app.rs:20]: Reference: https://support.stockcharts.com/doku.php?id=sharpcharts:indicator_past_data

Comment (leesiongchan) [reply src/app.rs:20]: Yea we can do that.

File: src/stock.rs [hunk 4]
```diff
@@ -14,8 +15,14 @@ impl Stock {
     pub async fn load_historical_prices(
         &mut self,
         time_frame: TimeFrame,
+        end_date: DateTime<Utc>,
     ) -> Result<(), Box<dyn Error>> {
-        self.bars = history::retrieve_interval(self.symbol.as_str(), time_frame.interval()).await?;
+        self.bars = if let Some(duration) = time_frame.duration() {
+            history::retrieve_range(self.symbol.as_str(), end_date - duration, Some(end_date))
```
Comment (teohhanhui) [inline src/stock.rs:21]: We see here that `retrieve_range` is designed with an optional end date in mind. So yeah, that kinda corroborates my earlier point about using a start date instead... :laughing:

Comment (teohhanhui) [reply src/stock.rs:21]: What we can do is not have a start date at first, but only calculate it the first time the user starts panning to the left.

File: src/main.rs [hunk 5]
```diff
@@ -210,46 +236,48 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
                         && area.bottom() >= row
                 }) {
-                    Some((UiTarget::StockSymbol, _)) | Some((UiTarget::StockName, _)) => {
-                        stock_symbol_input_state.active = !stock_symbol_input_state.active;
-                        time_frame_menu_state.active = false;
-                    }
-                    Some((UiTarget::TimeFrame, _)) => {
-                        time_frame_menu_state.active = !time_frame_menu_state.active;
-                        stock_symbol_input_state.active = false;
+                Some((UiTarget::StockSymbol, _)) | Some((UiTarget::StockName, _)) => {
```
Comment (teohhanhui) [inline src/main.rs:249]: Omg this indentation is so weird... :cry:

Comment (teohhanhui) [reply src/main.rs:249]: Is this a change in `rustfmt`? Hmm... I'm using stable Rust.

File: src/ui.rs (hunk 2 above)
Comment (teohhanhui) [reply src/ui.rs:91]: `rustc` bug? Haha... Are you using nightly? I'm on stable...

File: src/app.rs (hunk 1 above)
Comment (teohhanhui) [reply src/app.rs:174]: Oh, I think it's correct. Consider that trading days don't even match with calendar months in the first place. For example, no trading on weekends and public holidays.

File: src/main.rs (hunk 5 above)
Comment (leesiongchan) [reply src/main.rs:249]: Me `stable 1.43.1`, let me check again.

File: src/ui.rs (hunk 2 above)
Comment (leesiongchan) [reply src/ui.rs:91]: Previously you also have the `..`, just curious is it needed?

Comment (teohhanhui) [reply src/ui.rs:91]: Previously I wasn't using `ui_state` here... So no, it's not needed anymore if you use everything.

**Comment by teohhanhui** [comment]
Thanks @leesiongchan! :tada:

**Pull request merged**